File: app/src/main/java/com/byd/dashcast/util/PackagePseudonymizer.java
package com.byd.dashcast.util;

import java.nio.charset.StandardCharsets;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

/** Creates stable per-installation package markers without exposing a global dictionary hash. */
public final class PackagePseudonymizer {

    private static final char[] HEX = "0123456789abcdef".toCharArray();
    private static final int MARKER_BYTES = 8;

    private PackagePseudonymizer() {}

    /** Returns the first 64 bits of HMAC-SHA-256 as lower-case hexadecimal. */
    public static String marker(byte[] installationKey, String packageName) {
        if (installationKey == null || installationKey.length < 16) {
            throw new IllegalArgumentException("installationKey must contain at least 128 bits");
        }
        if (packageName == null) throw new IllegalArgumentException("packageName required");
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(installationKey, "HmacSHA256"));
            byte[] digest = mac.doFinal(packageName.getBytes(StandardCharsets.UTF_8));
            char[] marker = new char[MARKER_BYTES * 2];
            for (int i = 0; i < MARKER_BYTES; i++) {
                int value = digest[i] & 0xff;
                marker[i * 2] = HEX[value >>> 4];
                marker[i * 2 + 1] = HEX[value & 0x0f];
            }
            return new String(marker);
        } catch (java.security.GeneralSecurityException e) {
            throw new IllegalStateException("HmacSHA256 unavailable", e);
        }
    }
}

File: app/src/main/java/com/byd/dashcast/util/PackagePseudonymizer.kt
package com.byd.dashcast.util

import java.nio.charset.StandardCharsets
import java.security.GeneralSecurityException
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/** Creates stable per-installation package markers without exposing a global dictionary hash. */
object PackagePseudonymizer {

    private val HEX = "0123456789abcdef".toCharArray()
    private const val MARKER_BYTES = 8

    /** Returns the first 64 bits of HMAC-SHA-256 as lower-case hexadecimal. */
    @JvmStatic
    fun marker(installationKey: ByteArray?, packageName: String?): String {
        require(installationKey != null && installationKey.size >= 16) {
            "installationKey must contain at least 128 bits"
        }
        requireNotNull(packageName) { "packageName required" }
        return try {
            val mac = Mac.getInstance("HmacSHA256")
            mac.init(SecretKeySpec(installationKey, "HmacSHA256"))
            val digest = mac.doFinal(packageName.toByteArray(StandardCharsets.UTF_8))
            val marker = CharArray(MARKER_BYTES * 2)
            for (i in 0 until MARKER_BYTES) {
                val value = digest[i].toInt() and 0xff
                marker[i * 2] = HEX[value ushr 4]
                marker[i * 2 + 1] = HEX[value and 0x0f]
            }
            String(marker)
        } catch (e: GeneralSecurityException) {
            throw IllegalStateException("HmacSHA256 unavailable", e)
        }
    }
}
